transports/janus_websockets.c

```diff
@@ -781,52 +774,45 @@ gboolean janus_websockets_is_admin_api_enabled(void) {
 	return ws_admin_api_enabled;
 }
 
-int janus_websockets_send_message(void *transport, void *request_id, gboolean admin, json_t *message) {
+int janus_websockets_send_message(janus_transport_session *transport, void *request_id, gboolean admin, json_t *message) {
 	if(message == NULL)
 		return -1;
-	if(transport == NULL) {
+	if(transport == NULL || g_atomic_int_get(&transport->destroyed)) {
 		json_decref(message);
 		return -1;
 	}
-	/* Make sure this is not related to a closed /freed WebSocket session */
-	janus_mutex_lock(&old_wss_mutex);
-	janus_websockets_client *client = (janus_websockets_client *)transport;
-	if(g_list_find(old_wss, client) != NULL || !client->wsi) {
+	janus_mutex_lock(&transport->mutex);
+	janus_websockets_client *client = (janus_websockets_client *)transport->transport_p;
+	if(!client) {
 		json_decref(message);
-		message = NULL;
-		transport = NULL;
-		janus_mutex_unlock(&old_wss_mutex);
+		janus_mutex_unlock(&transport->mutex);
 		return -1;
 	}
-	janus_mutex_lock(&client->mutex);
 	/* Convert to string and enqueue */
 	char *payload = json_dumps(message, json_format);
 	g_async_queue_push(client->messages, payload);
 	lws_callback_on_writable(client->wsi);
-	janus_mutex_unlock(&client->mutex);
-	janus_mutex_unlock(&old_wss_mutex);
+	janus_mutex_unlock(&transport->mutex);
 	json_decref(message);
 	return 0;
 }
 
-void janus_websockets_session_created(void *transport, guint64 session_id) {
+void janus_websockets_session_created(janus_transport_session *transport, guint64 session_id) {
 	/* We don't care */
 }
 
-void janus_websockets_session_over(void *transport, guint64 session_id, gboolean timeout) {
-	if(transport == NULL || !timeout)
+void janus_websockets_session_over(janus_transport_session *transport, guint64 session_id, gboolean timeout) {
+	if(!timeout || transport == NULL || g_atomic_int_get(&transport->destroyed))
 		return;
+	janus_mutex_lock(&transport->mutex);
 	/* We only care if it's a timeout: if so, close the connection */
-	janus_websockets_client *client = (janus_websockets_client *)transport;
-	/* Make sure this is not related to a closed WebSocket session */
-	janus_mutex_lock(&old_wss_mutex);
-	if(g_list_find(old_wss, client) == NULL && client->wsi){
-		janus_mutex_lock(&client->mutex);
-		client->session_timeout = 1;
-		lws_callback_on_writable(client->wsi);
-		janus_mutex_unlock(&client->mutex);
+	janus_websockets_client *client = (janus_websockets_client *)transport->transport_p;
+	if(!client) {
+		janus_mutex_unlock(&transport->mutex);
+		return;
 	}
-	janus_mutex_unlock(&old_wss_mutex);
+	g_atomic_int_set(&client->session_timeout, 1);
+	lws_callback_on_writable(client->wsi);
```

Comment:
It looks suspicious that `&transport->mutex` is not getting unlocked in the case where `client` is non-null. Is this correct?

Reply:
Upon some experimentation, I think I can reliably demonstrate this to be problematic in the following case:

- At time T+0 seconds, I open a WS connection to Janus and create a session.
- At T+59, I send any message to Janus over that session.
- At T+60, Janus times out my session. This code runs and the transport session mutex remains locked.
- At T+61, Janus receives my message from two seconds prior on the timed-out session. The LWS_CALLBACK_SERVER_WRITEABLE code runs on the WS thread and tries to handle the message, but can't take this lock. The WS thread is then stuck waiting forever.

Let me know if you want me to hand you a test program that will simulate it. It seems 100% reliable to me.